File: app/src/main/java/com/teambrella/android/data/base/TeambrellaDataLoader.java
package com.teambrella.android.data.base;

import android.content.Context;
import android.net.Uri;

import com.google.gson.JsonObject;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.ui.TeambrellaUser;

import io.reactivex.Notification;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.observables.ConnectableObservable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;

/**
 * Teambrella data loader
 */
public class TeambrellaDataLoader {

    private ConnectableObservable<Notification<JsonObject>> mConnectableObservable;
    private PublishSubject<Notification<JsonObject>> mPublisher = PublishSubject.create();
    private TeambrellaServer mServer;


    public TeambrellaDataLoader(Context context) {
        mConnectableObservable = mPublisher.replay(1);
        mConnectableObservable.connect();
        mServer = new TeambrellaServer(context, TeambrellaUser.get(context).getPrivateKey());
    }


    public Observable<Notification<JsonObject>> getObservable() {
        return mConnectableObservable;
    }


    public void load(Uri uri, JsonObject data) {
        mServer.requestObservable(uri, data)
                .map(jsonObject -> {
                    if (jsonObject != null) {
                        jsonObject.get(TeambrellaModel.ATTR_STATUS).getAsJsonObject().addProperty(TeambrellaModel.ATTR_STATUS_URI, uri.toString());
                    }
                    return jsonObject;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onNext, this::onError, this::onComplete);
    }


    private void onNext(JsonObject data) {
        mPublisher.onNext(Notification.createOnNext(data));
    }

    private void onError(Throwable throwable) {
        mPublisher.onNext(Notification.createOnError(throwable));
    }

    private void onComplete() {
        // nothing to do
    }


}

File: app/src/main/java/com/teambrella/android/data/base/TeambrellaDataLoader.kt
package com.teambrella.android.data.base

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import android.net.Uri
import com.google.gson.JsonObject
import com.teambrella.android.api.server.TeambrellaServer
import com.teambrella.android.api.status
import com.teambrella.android.api.uri
import com.teambrella.android.dagger.Dependencies
import io.reactivex.Notification
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Named

class TeambrellaDataLoader {

    @Inject
    @field:Named(Dependencies.TEAMBRELLA_SERVER)
    lateinit var server: TeambrellaServer


    val observable: LiveData<Notification<JsonObject>> = MutableLiveData()


    fun load(uri: Uri, data: JsonObject? = null) {
        server.requestObservable(uri, data)?.map { response: JsonObject ->
            response.status?.uri = uri.toString()
            return@map response
        }?.subscribeOn(Schedulers.io())
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeAutoDispose(this::onNext, this::onError, this::onComplete)

    }

    internal fun update(updater: (JsonObject?) -> Boolean) {
        observable.value?.let { _notification ->
            var updated = false;
            if (_notification.isOnNext) {
                updated = updater(_notification.value)
            }
            if (updated) {
                onNext(_notification.value!!)
            }
        }
    }

    private fun onNext(item: JsonObject) {
        (observable as MutableLiveData).postValue(Notification.createOnNext<JsonObject>(item))
    }

    private fun onError(throwable: Throwable) {
        (observable as MutableLiveData).postValue(Notification.createOnError(throwable))
    }

    private fun onComplete() {

    }

}

fun <T> Observable<T>.subscribeAutoDispose(onNext: (T) -> Unit, onError: (Throwable) -> Unit,
                                           onComplete: () -> Unit) {
    lateinit var disposable: Disposable
    fun dispose() {
        disposable.dispose()
    }
    disposable = this.subscribe({
        dispose()
        onNext(it)
    }, {
        dispose()
        onError(it)
    }, {
        dispose()
        onComplete()
    })
}